RPC: Add API version to context response

jstarry opened the pull request:
#### Problem
When new API params are supported, it's not possible for clients to know if the new params are being ignored or not unless they first check the version of the RPC node they are connected to. Since RPC nodes are often used in pools, it's therefore quite difficult from the client to know what api version handled its request.

#### Summary of Changes
- Add the API version (solana software version) to the RPC response context
- Since subscriptions are long lived, it's not necessary to return the api version in each response so I've created new types for notification context responses.

```sh
curl 'http://localhost:8899' -X POST  -H 'content-type: application/json' --data-raw '{"method":"getAccountInfo","jsonrpc":"2.0","params":["DpVdRMmuvYZj3n4nCTJ7hufYi4odz2uNEsnfbH9vQ2sX"], "id":""}' | jq .                              
```
```json
{
  "jsonrpc": "2.0",
  "result": {
    "context": {
      "apiVersion": "1.11.0",
      "slot": 266
    },
    "value": {
      "data": "",
      "executable": false,
      "lamports": 499998670000,
      "owner": "11111111111111111111111111111111",
      "rentEpoch": 0
    }
  },
  "id": ""
}
```

Fixes #
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

CriesofCarrots reviewed (commented):
This looks good to me, but it's definitely breaking for rust pubsub_client users (as per the downstream-anchor CI failure). Maybe we're okay with that (and we just ignore that CI job until this is released and anchor is updated), but an alternative would be to revert the changes in pubsubs_client.rs, such that the client still expects RpcResponseContext. Can still save the wire space by sending the abbreviated struct, and api_version will always be populated as None on the client side.

jstarry commented:
> This looks good to me, but it's definitely breaking for rust pubsub_client users (as per the downstream-anchor CI failure). Maybe we're okay with that (and we just ignore that CI job until this is released and anchor is updated, but an alternative would be to revert the changes in pubsubs_client.rs, such that the client still expects RpcResponseContext. Can still save the wire space by sending the abbreviated struct, and api_version will always be populated as None on the client side.

Thanks for pointing that out and I'm happy the downstream test caught this. I didn't intend to break the API, I've reverted the pubsub client changes as you suggested and added some conversion logic to go from a notification type to a generic response type without an api version.

CriesofCarrots reviewed (approved):
Thanks, this lgtm!
